app/src/org/commcare/android/resource/AndroidResourceManager.java

```diff
@@ -0,0 +1,267 @@
+            UnresolvedResourceException,
+            InstallCancelledException {
+        tempUpgradeTable.destroy();
+        loadProfile(tempUpgradeTable, profileRef);
+        Resource tempProfile =
+                tempUpgradeTable.getResourceWithId(CommCarePlatform.APP_PROFILE_RESOURCE_ID);
+
+        if (tempProfile != null && tempProfile.isNewer(upgradeProfile)) {
+            upgradeTable.destroy();
+            tempUpgradeTable.copyToTable(upgradeTable);
+        }
+
+        tempUpgradeTable.destroy();
+    }
+
+    /**
+     * Set listeners and checkers that enable communication between low-level
+     * resource installation and top-level app update/installation process.
+     *
+     * @param tableListener  allows resource table to report its progress to the
+     *                       launching process
+     * @param cancelCheckker allows resource installers to check if the
+     *                       launching process was cancelled
+     */
+    @Override
+    public void setUpgradeListeners(TableStateListener tableListener,
+                                    InstallCancelled cancelCheckker) {
+        super.setUpgradeListeners(tableListener, cancelCheckker);
+
+        tempUpgradeTable.setStateListener(tableListener);
+        tempUpgradeTable.setInstallCancellationChecker(cancelCheckker);
+    }
+
+    /**
+     * Save upgrade stats if the upgrade was cancelled and wasn't complete at
+     * that time.
+     */
+    public void upgradeCancelled() {
+        if (!isUpgradeTableStaged()) {
+            UpdateStatPersistence.saveStatsPersistently(app, updateStats);
+        } else {
+            Log.i(TAG, "Upgrade cancelled, but already finished with these stats");
+            Log.i(TAG, updateStats.toString());
+        }
+    }
+
+    public void incrementUpdateAttempts() {
+        updateStats.registerStagingAttempt();
+    }
+
+    /**
+     * Log update failure that occurs while trying to install the staged update table
+     */
+    public void recordUpdateInstallFailure(Exception exception) {
+        updateStats.registerUpdateException(exception);
+    }
+
+    public void recordUpdateInstallFailure(AppInstallStatus result) {
+        updateStats.registerUpdateException(new Exception(result.toString()));
+    }
+
+    /**
+     * Clear update table, log failure with update stats,
+     * and, if appropriate, schedule a update retry
+     *
+     * @param exception Log error with update statistics
+     * @param context   Used for showing pinned notification of update task retry
+     */
+    public void processUpdateFailure(Exception exception, Context context) {
+        updateStats.registerUpdateException(exception);
+
+        upgradeTable.clear();
+
+        retryUpdateOrGiveUp(context);
+    }
+
+    public void processUpdateFailure(AppInstallStatus result, Context ctx) {
+        updateStats.registerUpdateException(new Exception(result.toString()));
+
+        boolean reusePartialTable =
```

Comment:
Not really sure. I did deviate from the old way of determining this: https://github.com/dimagi/commcare-odk/blob/master/app/src/org/commcare/dalvik/activities/CommCareSetupActivity.java#L482-L522.
Probably a good thing to go over with @ctsims